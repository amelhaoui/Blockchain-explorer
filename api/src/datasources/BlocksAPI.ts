import { RESTDataSource } from 'apollo-datasource-rest';
import { BLOCKCHAIN_URL } from 'config/constants';
class BlocksAPI extends RESTDataSource {
    constructor() {
        super();
        this.baseURL = BLOCKCHAIN_URL;
    }

    async getBlocks(time: string): Promise<unknown> {
        let timeMillis = Date.now();
        if (time) {
            timeMillis = parseInt(time);
        }
        return this.get(`blocks/${timeMillis}?format=json`);
    }

    async getBlock(hash: string): Promise<unknown> {
        return this.get(`rawblock/${encodeURIComponent(hash)}`);
    }
}

export default BlocksAPI;