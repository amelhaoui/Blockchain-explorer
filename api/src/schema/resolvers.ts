import { Resolvers, BlockSummary, Maybe} from './__generated/resolvers-types';

const resolvers : Resolvers = {
    Query: {
        blocks(_parent, {time, offset = 0, limit = 10}, {dataSources}):  Array<BlockSummary> {
           return dataSources.BlocksAPI.getBlocks(time).then((value: Array<Maybe<BlockSummary>>) =>  value.slice(offset, offset + limit));
        },
        block(_parent, {hash}, {dataSources}) {
          return dataSources.BlocksAPI.getBlock(hash);
        }
    }
}

export default resolvers;