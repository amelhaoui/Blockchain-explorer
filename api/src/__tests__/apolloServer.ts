import { ApolloServer, gql } from "apollo-server-express";
import BlocksAPI from "datasources/BlocksAPI";
import resolvers from "schema/resolvers";
import typeDefs from "schema/schema";

// copied from front
const GET_BLOCKS = gql`
  query GetBlocks($time: String!, $offset: Int!, $limit: Int!) {
    blocks(time: $time, offset: $offset, limit: $limit) {
      hash
      height
      time
    }
  }
`;

describe('API integration', () => {
    it('fetches 10 blocks', async () => {
        // create a test server to test against, using our production typeDefs,
        // resolvers, and dataSources.
        const server = new ApolloServer({
            typeDefs,
            resolvers,
            dataSources: () => ({ BlocksAPI: new BlocksAPI() }),
        });
        const res = await server.executeOperation({ query: GET_BLOCKS, variables: { time: "1633272354897", offset: 0, limit: 10} });
        expect(res).toMatchSnapshot();
    });
});