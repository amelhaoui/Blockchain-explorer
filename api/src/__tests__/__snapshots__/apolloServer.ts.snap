// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API integration fetches 10 blocks 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "hash": "0000000000000000000dc7e9e2b6b460b09658ede1a7235377eb5cc6a12ea29b",
        "height": 703376,
        "time": "1633272049",
      },
      Object {
        "hash": "00000000000000000001d3d60facac6a44f58ae34e8b4a064f16be8432388f22",
        "height": 703375,
        "time": "1633272021",
      },
      Object {
        "hash": "000000000000000000099cd7c12e7b4df8e19d2469694c66be038057c8e5639b",
        "height": 703374,
        "time": "1633271921",
      },
      Object {
        "hash": "0000000000000000000949c5edc618671b1df1255c5a477428e51fe4a360ffc8",
        "height": 703373,
        "time": "1633271415",
      },
      Object {
        "hash": "0000000000000000000276388ba18ce921d7d825863c4bb6a3b607e1dafb6e7e",
        "height": 703372,
        "time": "1633270826",
      },
      Object {
        "hash": "000000000000000000095568fcb927e1b0e519ca185bb87f10cb76fd7df3730b",
        "height": 703371,
        "time": "1633268867",
      },
      Object {
        "hash": "0000000000000000000352358c54e848b2ed1bce333911c3645c95bbf583abe7",
        "height": 703370,
        "time": "1633267705",
      },
      Object {
        "hash": "000000000000000000070f9569c445e2de1addba69e1046aa5ce49329cabec01",
        "height": 703369,
        "time": "1633266582",
      },
      Object {
        "hash": "0000000000000000000b4272e298930d257f01bf561f51cf4d231aa50d003e90",
        "height": 703368,
        "time": "1633266400",
      },
      Object {
        "hash": "00000000000000000007cdf638de8885341536ab7f56fd20972b0c852c9a55e1",
        "height": 703367,
        "time": "1633266209",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
